crates/task-impls/src/quorum_proposal_recv/handlers.rs

```diff
@@ -151,11 +153,25 @@ pub(crate) async fn handle_quorum_proposal_recv<
 
     let view_number = proposal.data.view_number();
     let justify_qc = proposal.data.justify_qc.clone();
+    let proposal_block_number = proposal.data.block_header.block_number();
+    let proposal_epoch = TYPES::Epoch::new(epoch_from_block_number(
+        proposal_block_number,
+        validation_info.epoch_height,
+    ));
+    let justify_qc_epoch = if validation_info.epoch_height != 0
+        && proposal_block_number % validation_info.epoch_height == 1
+    {
+        // if the proposal is for the first block in an epoch, the justify QC must be from the previous epoch
+        proposal_epoch - 1
+    } else {
+        // otherwise justify QC is from the same epoch
+        proposal_epoch
+    };
```

Comment:
is it worth making this a helper function? i.e. take a proposal and the previous block header and return the proposal's epoch number

Reply:
I guess I meant:

- the proposed block's epoch is already `epoch_from_block_number(proposal.data.block_header.block_number())` (straightforward)
- the preceding block's epoch is the check based on the block height modulo the epoch height, but we could make a function like `preceding_block_epoch(proposal, epoch_height)`

(maybe these could even be generic in a `HasBlockNumber` trait or something)

it's not that important though!